diff --git a/CMakeLists.txt b/CMakeLists.txt
index 676727f1..85af1083 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,7 +19,7 @@
 #       start libevent.sln
 #
 
-cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
 
 if (POLICY CMP0054)
     cmake_policy(SET CMP0054 NEW)
@@ -1453,7 +1453,7 @@ endif()
 #
 
 set(EVENT_INSTALL_CMAKE_DIR
-    "${CMAKE_INSTALL_PREFIX}/lib/cmake/libevent")
+    "lib/cmake/libevent")
 
 export(PACKAGE libevent)
 
diff --git a/cmake/AddEventLibrary.cmake b/cmake/AddEventLibrary.cmake
index 04f5837e..95d98087 100644
--- a/cmake/AddEventLibrary.cmake
+++ b/cmake/AddEventLibrary.cmake
@@ -42,7 +42,7 @@ macro(export_install_target TYPE LIB_NAME OUTER_INCLUDES)
         install(TARGETS "${LIB_NAME}_${TYPE}"
             LIBRARY DESTINATION "lib" COMPONENT lib
             ARCHIVE DESTINATION "lib" COMPONENT lib
-            RUNTIME DESTINATION "lib" COMPONENT lib
+            RUNTIME DESTINATION "bin" COMPONENT bin
             COMPONENT dev
         )
     else()
@@ -69,7 +69,7 @@ macro(export_install_target TYPE LIB_NAME OUTER_INCLUDES)
             EXPORT LibeventTargets-${TYPE}
             LIBRARY DESTINATION "lib" COMPONENT lib
             ARCHIVE DESTINATION "lib" COMPONENT lib
-            RUNTIME DESTINATION "lib" COMPONENT lib
+            RUNTIME DESTINATION "bin" COMPONENT bin
             COMPONENT dev
         )
     endif()
diff --git a/cmake/LibeventConfig.cmake.in b/cmake/LibeventConfig.cmake.in
index 7b808c32..8f462c71 100644
--- a/cmake/LibeventConfig.cmake.in
+++ b/cmake/LibeventConfig.cmake.in
@@ -130,8 +130,7 @@ if(CONFIG_FOR_INSTALL_TREE)
         unset(_event_lib CACHE)
         find_library(_event_lib
                     NAMES "event_${_comp}"
-                    PATHS "${_INSTALL_PREFIX}/lib"
-                    NO_DEFAULT_PATH)
+                    PATHS "${_INSTALL_PREFIX}/lib")
         if(_event_lib)
             list(APPEND LIBEVENT_LIBRARIES "libevent::${_comp}")
             set_case_insensitive_found(${_comp})
