diff --git a/absl/base/config.h b/absl/base/config.h
index 585485c..be69e70 100644
--- a/absl/base/config.h
+++ b/absl/base/config.h
@@ -550,7 +550,8 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
 
 #ifdef __has_include
 #if __has_include(<any>) && defined(__cplusplus) && __cplusplus >= 201703L && \
-    !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE
+    !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE && \
+    !defined(_MSC_VER)
 #define ABSL_HAVE_STD_ANY 1
 #endif
 #endif
@@ -564,7 +565,8 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
 
 #ifdef __has_include
 #if __has_include(<optional>) && defined(__cplusplus) && \
-    __cplusplus >= 201703L && !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE
+    __cplusplus >= 201703L && !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE && \
+    !defined(_MSC_VER)
 #define ABSL_HAVE_STD_OPTIONAL 1
 #endif
 #endif
@@ -578,7 +580,8 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
 
 #ifdef __has_include
 #if __has_include(<variant>) && defined(__cplusplus) && \
-    __cplusplus >= 201703L && !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE
+    __cplusplus >= 201703L && !ABSL_INTERNAL_APPLE_CXX17_TYPES_UNAVAILABLE && \
+    !defined(_MSC_VER)
 #define ABSL_HAVE_STD_VARIANT 1
 #endif
 #endif
@@ -592,7 +595,8 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
 
 #ifdef __has_include
 #if __has_include(<string_view>) && defined(__cplusplus) && \
-    __cplusplus >= 201703L
+    __cplusplus >= 201703L && \
+    !defined(_MSC_VER)
 #define ABSL_HAVE_STD_STRING_VIEW 1
 #endif
 #endif
@@ -608,9 +612,9 @@ static_assert(ABSL_INTERNAL_INLINE_NAMESPACE_STR[0] != 'h' ||
     ((defined(_MSVC_LANG) && _MSVC_LANG > 201402) || \
      (defined(__cplusplus) && __cplusplus > 201402))
 // #define ABSL_HAVE_STD_ANY 1
-#define ABSL_HAVE_STD_OPTIONAL 1
-#define ABSL_HAVE_STD_VARIANT 1
-#define ABSL_HAVE_STD_STRING_VIEW 1
+// #define ABSL_HAVE_STD_OPTIONAL 1
+// #define ABSL_HAVE_STD_VARIANT 1
+// #define ABSL_HAVE_STD_STRING_VIEW 1
 #endif
 
 // ABSL_USES_STD_ANY
