diff --git a/third_party/upb/upb/text_encode.c b/third_party/upb/upb/text_encode.c
index 028cc29..bf23050 100644
--- a/third_party/upb/upb/text_encode.c
+++ b/third_party/upb/upb/text_encode.c
@@ -245,7 +245,7 @@ static void txtenc_map(txtenc *e, const upb_map *map, const upb_fielddef *f) {
   }
 }

-#define CHK(x) do { if (!(x)) { return false; } } while(0)
+#define CHK(x) do { if (!(x)) { return NULL; } } while(0)

 static const char *txtenc_parsevarint(const char *ptr, const char *limit,
                                       uint64_t *val) {
