diff --git a/CMakeLists.txt b/CMakeLists.txt
index 25990a5..1d39320 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -4198,6 +4198,7 @@ add_library(upb
   src/core/ext/upb-generated/google/protobuf/descriptor.upb.c
   src/core/ext/upbdefs-generated/google/protobuf/descriptor.upbdefs.c
 )
+set_target_properties(upb PROPERTIES OUTPUT_NAME "grpc_upb")
 
 set_target_properties(upb PROPERTIES
   VERSION ${gRPC_CORE_VERSION}
diff --git a/third_party/upb/upb/text_encode.c b/third_party/upb/upb/text_encode.c
index 9612375..76b0bfb 100644
--- a/third_party/upb/upb/text_encode.c
+++ b/third_party/upb/upb/text_encode.c
@@ -294,7 +294,7 @@ static void txtenc_map(txtenc* e, const upb_Map* map, const upb_FieldDef* f) {
 #define CHK(x)      \
   do {              \
     if (!(x)) {     \
-      return false; \
+      return NULL;  \
     }               \
   } while (0)
 
