diff --git a/src/core/lib/gprpp/per_cpu.h b/src/core/lib/gprpp/per_cpu.h
index a4e30092..c43a7329 100644
--- a/src/core/lib/gprpp/per_cpu.h
+++ b/src/core/lib/gprpp/per_cpu.h
@@ -83,7 +83,7 @@ class PerCpu {
  public:
   // Options are not defaulted to try and force consideration of what the
   // options specify.
-  explicit PerCpu(PerCpuOptions options) : shards_(options.Shards()) {}
+  explicit PerCpu(PerCpuOptions options) : shards_(options.Shards()), data_{new T[options.Shards()]} {}
 
   T& this_cpu() { return data_[sharding_helper_.GetShardingBits() % shards_]; }
 
@@ -95,7 +95,7 @@ class PerCpu {
  private:
   PerCpuShardingHelper sharding_helper_;
   const size_t shards_;
-  std::unique_ptr<T[]> data_{new T[shards_]};
+  std::unique_ptr<T[]> data_;
 };
 
 }  // namespace grpc_core
diff --git a/third_party/upb/upb/text/encode.c b/third_party/upb/upb/text/encode.c
index e13d9f29..908f2332 100644
--- a/third_party/upb/upb/text/encode.c
+++ b/third_party/upb/upb/text/encode.c
@@ -378,7 +378,7 @@ static void txtenc_map(txtenc* e, const upb_Map* map, const upb_FieldDef* f) {
 #define CHK(x)      \
   do {              \
     if (!(x)) {     \
-      return false; \
+      return NULL;  \
     }               \
   } while (0)
 
diff --git a/third_party/upb/upb/wire/internal/decoder.h b/third_party/upb/upb/wire/internal/decoder.h
index 5ca1e1b2..ea3eb2cf 100644
--- a/third_party/upb/upb/wire/internal/decoder.h
+++ b/third_party/upb/upb/wire/internal/decoder.h
@@ -17,7 +17,7 @@
 #include "upb/message/internal/message.h"
 #include "upb/wire/decode.h"
 #include "upb/wire/eps_copy_input_stream.h"
-#include "utf8_range.h"
+#include "../utf8_range/utf8_range.h"
 
 // Must be last.
 #include "upb/port/def.inc"
