diff --git a/src/core/lib/gprpp/per_cpu.h b/src/core/lib/gprpp/per_cpu.h
index a4e3009..c43a732 100644
--- a/src/core/lib/gprpp/per_cpu.h
+++ b/src/core/lib/gprpp/per_cpu.h
@@ -83,7 +83,7 @@ class PerCpu {
  public:
   // Options are not defaulted to try and force consideration of what the
   // options specify.
-  explicit PerCpu(PerCpuOptions options) : shards_(options.Shards()) {}
+  explicit PerCpu(PerCpuOptions options) : shards_(options.Shards()), data_{new T[options.Shards()]} {}
 
   T& this_cpu() { return data_[sharding_helper_.GetShardingBits() % shards_]; }
 
@@ -95,7 +95,7 @@ class PerCpu {
  private:
   PerCpuShardingHelper sharding_helper_;
   const size_t shards_;
-  std::unique_ptr<T[]> data_{new T[shards_]};
+  std::unique_ptr<T[]> data_;
 };
 
 }  // namespace grpc_core
diff --git a/third_party/upb/upb/message/copy.c b/third_party/upb/upb/message/copy.c
index 114b92e..84e6f83 100644
--- a/third_party/upb/upb/message/copy.c
+++ b/third_party/upb/upb/message/copy.c
@@ -150,7 +150,7 @@ upb_Array* upb_Array_DeepClone(const upb_Array* array, upb_CType value_type,
   for (size_t i = 0; i < size; ++i) {
     upb_MessageValue val = upb_Array_Get(array, i);
     if (!upb_Clone_MessageValue(&val, value_type, sub, arena)) {
-      return false;
+      return NULL;
     }
     upb_Array_Set(cloned_array, i, val);
   }
diff --git a/third_party/upb/upb/text/encode.c b/third_party/upb/upb/text/encode.c
index 055d005..ca84083 100644
--- a/third_party/upb/upb/text/encode.c
+++ b/third_party/upb/upb/text/encode.c
@@ -385,7 +385,7 @@ static void txtenc_map(txtenc* e, const upb_Map* map, const upb_FieldDef* f) {
 #define CHK(x)      \
   do {              \
     if (!(x)) {     \
-      return false; \
+      return NULL;  \
     }               \
   } while (0)
 
diff --git a/third_party/upb/upb/wire/internal/decoder.h b/third_party/upb/upb/wire/internal/decoder.h
index 5ca1e1b..ea3eb2c 100644
--- a/third_party/upb/upb/wire/internal/decoder.h
+++ b/third_party/upb/upb/wire/internal/decoder.h
@@ -17,7 +17,7 @@
 #include "upb/message/internal/message.h"
 #include "upb/wire/decode.h"
 #include "upb/wire/eps_copy_input_stream.h"
-#include "utf8_range.h"
+#include "../utf8_range/utf8_range.h"
 
 // Must be last.
 #include "upb/port/def.inc"
