diff --git a/include/flatbuffers/base.h b/include/flatbuffers/base.h
index 1c19dde..f160e62 100644
--- a/include/flatbuffers/base.h
+++ b/include/flatbuffers/base.h
@@ -200,7 +200,7 @@ namespace flatbuffers {
 #endif
 
 #if (!defined(_MSC_VER) || _MSC_VER >= 1900) && \
-    (!defined(__GNUC__) || (__GNUC__ * 100 + __GNUC_MINOR__ >= 409)) || \
+    (!defined(__GNUC__) || (__GNUC__ * 100 + __GNUC_MINOR__ >= 408)) || \
     defined(__clang__)
   #define FLATBUFFERS_DEFAULT_DECLARATION
 #endif
