diff --git a/CMakeLists.txt b/CMakeLists.txt
index 53a4255..a1abb6f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -337,7 +337,7 @@ if(APPLE)
   add_definitions(-D__APPLE_USE_RFC_3542)
 endif()
 
-if(ENABLE_SHARED_LIB AND ENABLE_STATIC_LIB AND MSVC AND NOT STATIC_LIB_SUFFIX)
+if(ENABLE_SHARED_LIB AND ENABLE_STATIC_LIB AND NOT STATIC_LIB_SUFFIX)
   set(STATIC_LIB_SUFFIX "_static")
 endif()
 
@@ -355,7 +355,7 @@ if(BUILD_TESTING)
 endif()
 add_subdirectory(crypto)
 add_subdirectory(third-party)
-add_subdirectory(examples)
+# add_subdirectory(examples)
 
 
 string(TOUPPER "${CMAKE_BUILD_TYPE}" _build_type)
