diff --git a/src/alloc-override.c b/src/alloc-override.c
index 84a0d19..9066a1d 100644
--- a/src/alloc-override.c
+++ b/src/alloc-override.c
@@ -57,7 +57,7 @@ typedef struct mi_nothrow_s { int _tag; } mi_nothrow_t;
   // functions that are interposed (or the interposing does not work)
   #define MI_OSX_IS_INTERPOSED
 
-  mi_decl_externc static size_t mi_malloc_size_checked(void *p) {
+  static size_t mi_malloc_size_checked(void *p) {
     if (!mi_is_in_heap_region(p)) return 0;
     return mi_usable_size(p);
   }
diff --git a/src/alloc.c b/src/alloc.c
index 86453f1..b2ac4cc 100644
--- a/src/alloc.c
+++ b/src/alloc.c
@@ -876,10 +876,10 @@ but we call `exit` instead (i.e. not returning).
 #ifdef __cplusplus
 #include <new>
 static bool mi_try_new_handler(bool nothrow) {
-  #if defined(_MSC_VER) || (__cplusplus >= 201103L)
+  #if (defined(_MSC_VER) || (__cplusplus >= 201103L)) && !(defined(__GNUC__) && __GNUC__ < 5 && !defined(__clang__))
     std::new_handler h = std::get_new_handler();
   #else
-    std::new_handler h = std::set_new_handler();
+    std::new_handler h = std::set_new_handler(nullptr);
     std::set_new_handler(h);
   #endif
   if (h==NULL) {
