diff --git a/lib/CMakeLists.txt b/lib/CMakeLists.txt
index c6b9793..61f6060 100644
--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -100,6 +100,7 @@ endif()
 add_library(${PROJECT_NAME}::nghttp2 ALIAS ${LIB_SELECTED})
 
 include(CMakePackageConfigHelpers)
+include(GNUInstallDirs)
 set(INCLUDE_INSTALL_DIR include)
 file(MAKE_DIRECTORY "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/cmake/${PROJECT_NAME}")
 configure_package_config_file(
