diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4402af2..9dd2895 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -279,10 +279,28 @@ endif()
 
 if (ANDROID_PLATFORM)
     if (${ANDROID_STL} STREQUAL "c++_shared")
+        if (EXISTS "${ANDROID_NDK}/sources/cxx-stl/llvm-libc++/libs/${ANDROID_ABI}/libc++_shared.so")
         configure_file(
         "${ANDROID_NDK}/sources/cxx-stl/llvm-libc++/libs/${ANDROID_ABI}/libc++_shared.so"
         "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/libc++_shared.so"
         COPYONLY)
+        else()
+            if(CMAKE_ANDROID_ARCH_ABI STREQUAL armeabi-v7a)
+                set(TBB_ANDROID_TOOLCHAIN_NAME arm-linux-androideabi)
+            elseif(CMAKE_ANDROID_ARCH_ABI STREQUAL arm64-v8a)
+                set(TBB_ANDROID_TOOLCHAIN_NAME aarch64-linux-android)
+            elseif(CMAKE_ANDROID_ARCH_ABI STREQUAL x86)
+                set(TBB_ANDROID_TOOLCHAIN_NAME i686-linux-android)
+            elseif(CMAKE_ANDROID_ARCH_ABI STREQUAL x86_64)
+                set(TBB_ANDROID_TOOLCHAIN_NAME x86_64-linux-android)
+            else()
+            message(FATAL_ERROR "Invalid Android ABI: ${ANDROID_ABI}.")
+            endif()
+            configure_file(
+                "${ANDROID_TOOLCHAIN_ROOT}/sysroot/usr/lib/${TBB_ANDROID_TOOLCHAIN_NAME}/libc++_shared.so"
+                "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/libc++_shared.so"
+                COPYONLY)
+        endif()
     endif()
     # This custom target may be implemented without separate CMake script, but it requires
     # ADB(Android Debug Bridge) executable file availability, so to incapsulate this requirement
diff --git a/cmake/compilers/Clang.cmake b/cmake/compilers/Clang.cmake
index 69aa519..52d50dd 100644
--- a/cmake/compilers/Clang.cmake
+++ b/cmake/compilers/Clang.cmake
@@ -54,9 +54,9 @@ if (ANDROID_PLATFORM)
     set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} $<$<NOT:$<CONFIG:Debug>>:-D_FORTIFY_SOURCE=2>)
 endif()
 
-if (MINGW)
-    list(APPEND TBB_COMMON_COMPILE_FLAGS -U__STRICT_ANSI__)
-endif()
+# if (MINGW)
+#     list(APPEND TBB_COMMON_COMPILE_FLAGS -U__STRICT_ANSI__)
+# endif()
 
 set(TBB_IPO_COMPILE_FLAGS $<$<NOT:$<CONFIG:Debug>>:-flto>)
 set(TBB_IPO_LINK_FLAGS $<$<NOT:$<CONFIG:Debug>>:-flto>)
diff --git a/cmake/utils.cmake b/cmake/utils.cmake
index f74abfc..a090bac 100644
--- a/cmake/utils.cmake
+++ b/cmake/utils.cmake
@@ -18,7 +18,7 @@ macro(tbb_remove_compile_flag flag)
     set_property(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY COMPILE_OPTIONS ${_tbb_compile_options})
     unset(_tbb_compile_options)
     if (CMAKE_CXX_FLAGS)
-        string(REGEX REPLACE ${flag} "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
+        string(REGEX REPLACE "(^|[ \t\r\n]+)${flag}($|[ \t\r\n]+)" " " CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
     endif()
 endmacro()
 
