diff --git a/xxhash.h b/xxhash.h
index 2d56d23..409c9c6 100644
--- a/xxhash.h
+++ b/xxhash.h
@@ -639,7 +639,7 @@ struct XXH64_state_s {
    XXH64_hash_t reserved64;  /* never read nor write, might be removed in a future version */
 };   /* typedef'd to XXH64_state_t */
 
-#if defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)   /* C11+ */
+#if !defined(_MSC_VER) && defined (__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)   /* C11+ */
 #  include <stdalign.h>
 #  define XXH_ALIGN(n)      alignas(n)
 #elif defined(__GNUC__)
