diff --git a/cmake/patch-imported-config.cmake b/cmake/patch-imported-config.cmake
index ec68d740..a2d22ed3 100644
--- a/cmake/patch-imported-config.cmake
+++ b/cmake/patch-imported-config.cmake
@@ -21,8 +21,11 @@ if(TARGET c-ares::cares)
 endif()
 
 # curl targets
-if(TARGET CURL::libcurl)
-  project_build_tools_patch_default_imported_config(CURL::libcurl)
+if(TARGET CURL::libcurl
+   OR TARGET CURL::libcurl_static
+   OR TARGET CURL::libcurl_shared)
+  project_build_tools_patch_default_imported_config(
+    CURL::libcurl CURL::libcurl_static CURL::libcurl_shared)
 endif()
 
 # abseil targets
diff --git a/cmake/tools.cmake b/cmake/tools.cmake
index ee191121..fa66342a 100644
--- a/cmake/tools.cmake
+++ b/cmake/tools.cmake
@@ -146,6 +146,10 @@ function(project_build_tools_patch_default_imported_config)
       if(NOT IS_IMPORTED_TARGET)
         continue()
       endif()
+      get_target_property(IS_ALIAS_TARGET ${TARGET_NAME} ALIASED_TARGET)
+      if(IS_ALIAS_TARGET)
+        continue()
+      endif()
 
       if(CMAKE_VERSION VERSION_LESS "3.19.0")
         get_target_property(TARGET_TYPE_NAME ${TARGET_NAME} TYPE)
diff --git a/exporters/prometheus/src/exporter_utils.cc b/exporters/prometheus/src/exporter_utils.cc
index 4b1ee054..90583b0f 100644
--- a/exporters/prometheus/src/exporter_utils.cc
+++ b/exporters/prometheus/src/exporter_utils.cc
@@ -1,6 +1,7 @@
 // Copyright The OpenTelemetry Authors
 // SPDX-License-Identifier: Apache-2.0
 
+#include <algorithm>
 #include <limits>
 #include <regex>
 #include <sstream>
@@ -277,11 +278,7 @@ std::string PrometheusExporterUtils::SanitizeNames(std::string name)
   return name;
 }
 
-std::regex INVALID_CHARACTERS_PATTERN("[^a-zA-Z0-9]");
 std::regex CHARACTERS_BETWEEN_BRACES_PATTERN("\\{(.*?)\\}");
-std::regex SANITIZE_LEADING_UNDERSCORES("^_+");
-std::regex SANITIZE_TRAILING_UNDERSCORES("_+$");
-std::regex SANITIZE_CONSECUTIVE_UNDERSCORES("[_]{2,}");
 
 std::string PrometheusExporterUtils::GetEquivalentPrometheusUnit(
     const std::string &raw_metric_unit_name)
@@ -357,7 +354,7 @@ std::string PrometheusExporterUtils::GetPrometheusPerUnit(const std::string &per
 
 std::string PrometheusExporterUtils::RemoveUnitPortionInBraces(const std::string &unit)
 {
-  return std::regex_replace(unit, CHARACTERS_BETWEEN_BRACES_PATTERN, "");
+  return std::regex_replace(unit, CHARACTERS_BETWEEN_BRACES_PATTERN, std::string(""));
 }
 
 std::string PrometheusExporterUtils::ConvertRateExpressedToPrometheusUnit(
@@ -386,10 +383,65 @@ std::string PrometheusExporterUtils::ConvertRateExpressedToPrometheusUnit(
 
 std::string PrometheusExporterUtils::CleanUpString(const std::string &str)
 {
-  std::string cleaned_string = std::regex_replace(str, INVALID_CHARACTERS_PATTERN, "_");
-  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_CONSECUTIVE_UNDERSCORES, "_");
-  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_TRAILING_UNDERSCORES, "");
-  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_LEADING_UNDERSCORES, "");
+  std::string cleaned_string = str;
+  std::transform(cleaned_string.begin(), cleaned_string.end(), cleaned_string.begin(),
+                 [](const char c) {
+                   if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))
+                   {
+                     return c;
+                   }
+                   return '_';
+                 });
+
+  std::string::size_type trim_start = 0;
+  std::string::size_type trim_end   = 0;
+  bool previous_underscore          = false;
+  for (std::string::size_type i = 0; i < cleaned_string.size(); ++i)
+  {
+    if (cleaned_string[i] == '_')
+    {
+      if (previous_underscore)
+      {
+        continue;
+      }
+
+      previous_underscore = true;
+    }
+    else
+    {
+      previous_underscore = false;
+    }
+
+    if (trim_end != i)
+    {
+      cleaned_string[trim_end] = cleaned_string[i];
+    }
+    ++trim_end;
+  }
+
+  if (trim_start >= trim_end)
+  {
+    return "";
+  }
+
+  while (trim_end > 0 && cleaned_string[trim_end - 1] == '_')
+  {
+    --trim_end;
+  }
+  while (trim_start < trim_end && cleaned_string[trim_start] == '_')
+  {
+    ++trim_start;
+  }
+
+  // All characters are underscore
+  if (trim_start >= trim_end)
+  {
+    return "_";
+  }
+  if (0 != trim_start || cleaned_string.size() != trim_end)
+  {
+    return cleaned_string.substr(trim_start, trim_end - trim_start);
+  }
 
   return cleaned_string;
 }
