diff --git a/sdk/include/opentelemetry/sdk/metrics/sync_instruments.h b/sdk/include/opentelemetry/sdk/metrics/sync_instruments.h
index b754c19..0307faf 100644
--- a/sdk/include/opentelemetry/sdk/metrics/sync_instruments.h
+++ b/sdk/include/opentelemetry/sdk/metrics/sync_instruments.h
@@ -180,9 +180,9 @@ class BoundUpDownCounter final : public BoundSynchronousInstrument<T>,
 {
 
 public:
-  BoundUpDownCounter<T>() = default;
+  BoundUpDownCounter() = default;
 
-  BoundUpDownCounter<T>(nostd::string_view name,
+  BoundUpDownCounter(nostd::string_view name,
                         nostd::string_view description,
                         nostd::string_view unit,
                         bool enabled)
