diff --git a/ext/CMakeLists.txt b/ext/CMakeLists.txt
index a2baa47..034328e 100644
--- a/ext/CMakeLists.txt
+++ b/ext/CMakeLists.txt
@@ -5,6 +5,7 @@ target_include_directories(
             "$<INSTALL_INTERFACE:include>")
 
 set_target_properties(opentelemetry_ext PROPERTIES EXPORT_NAME "ext")
+target_link_libraries(opentelemetry_ext INTERFACE opentelemetry_api)
 
 install(
   TARGETS opentelemetry_ext
