diff --git a/exporters/elasticsearch/CMakeLists.txt b/exporters/elasticsearch/CMakeLists.txt
index 0c303db..1bd6716 100644
--- a/exporters/elasticsearch/CMakeLists.txt
+++ b/exporters/elasticsearch/CMakeLists.txt
@@ -3,6 +3,9 @@ include_directories(${CMAKE_SOURCE_DIR}/ext/include)
 
 add_library(opentelemetry_exporter_elasticsearch_logs src/es_log_exporter.cc)
 
+target_link_libraries(opentelemetry_exporter_elasticsearch_logs
+                      PUBLIC opentelemetry_trace http_client_curl)
+
 if(BUILD_TESTING)
   add_executable(es_log_exporter_test test/es_log_exporter_test.cc)
 
diff --git a/exporters/otlp/CMakeLists.txt b/exporters/otlp/CMakeLists.txt
index 0529436..a043d87 100644
--- a/exporters/otlp/CMakeLists.txt
+++ b/exporters/otlp/CMakeLists.txt
@@ -4,8 +4,10 @@ add_library(opentelemetry_exporter_otprotocol src/recordable.cc
 set_target_properties(opentelemetry_exporter_otprotocol
                       PROPERTIES EXPORT_NAME otlp_exporter)
 
-target_link_libraries(opentelemetry_exporter_otprotocol
-                      PUBLIC opentelemetry_trace opentelemetry_proto)
+target_link_libraries(
+  opentelemetry_exporter_otprotocol
+  PUBLIC opentelemetry_trace opentelemetry_proto protobuf::libprotobuf
+         gRPC::grpc++)
 
 install(
   TARGETS opentelemetry_exporter_otprotocol
diff --git a/exporters/zipkin/CMakeLists.txt b/exporters/zipkin/CMakeLists.txt
index 5d4fe8d..eefcf2c 100644
--- a/exporters/zipkin/CMakeLists.txt
+++ b/exporters/zipkin/CMakeLists.txt
@@ -20,6 +20,9 @@ endif()
 add_library(opentelemetry_exporter_zipkin_trace src/zipkin_exporter.cc
                                                 src/recordable.cc)
 
+target_link_libraries(opentelemetry_exporter_zipkin_trace
+                      PUBLIC opentelemetry_trace http_client_curl)
+
 install(
   TARGETS opentelemetry_exporter_zipkin_trace
   EXPORT "${PROJECT_NAME}-target"
diff --git a/sdk/src/common/core.cc b/sdk/src/common/core.cc
index a8db9a0..280ef70 100644
--- a/sdk/src/common/core.cc
+++ b/sdk/src/common/core.cc
@@ -2,6 +2,19 @@
 #include "opentelemetry/version.h"
 
 #if defined(HAVE_ABSEIL)
+// ================ __cdecl ================
+#if defined(__GNUC__) || defined(__GNUG__)
+#ifndef __cdecl
+// see https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html
+// Intel x86 architecture specific calling conventions
+#ifdef _M_IX86
+#define __cdecl __attribute__((__cdecl__))
+#else
+#define __cdecl
+#endif
+#endif
+#endif
+// ---------------- __cdecl ----------------
 namespace absl
 {
 namespace variant_internal
