diff --git a/cmake/patch-imported-config.cmake b/cmake/patch-imported-config.cmake
index ec68d740..a2d22ed3 100644
--- a/cmake/patch-imported-config.cmake
+++ b/cmake/patch-imported-config.cmake
@@ -21,8 +21,11 @@ if(TARGET c-ares::cares)
 endif()
 
 # curl targets
-if(TARGET CURL::libcurl)
-  project_build_tools_patch_default_imported_config(CURL::libcurl)
+if(TARGET CURL::libcurl
+   OR TARGET CURL::libcurl_static
+   OR TARGET CURL::libcurl_shared)
+  project_build_tools_patch_default_imported_config(
+    CURL::libcurl CURL::libcurl_static CURL::libcurl_shared)
 endif()
 
 # abseil targets
diff --git a/cmake/tools.cmake b/cmake/tools.cmake
index ee191121..fa66342a 100644
--- a/cmake/tools.cmake
+++ b/cmake/tools.cmake
@@ -146,6 +146,10 @@ function(project_build_tools_patch_default_imported_config)
       if(NOT IS_IMPORTED_TARGET)
         continue()
       endif()
+      get_target_property(IS_ALIAS_TARGET ${TARGET_NAME} ALIASED_TARGET)
+      if(IS_ALIAS_TARGET)
+        continue()
+      endif()
 
       if(CMAKE_VERSION VERSION_LESS "3.19.0")
         get_target_property(TARGET_TYPE_NAME ${TARGET_NAME} TYPE)
