diff --git a/exporters/otlp/src/otlp_file_client.cc b/exporters/otlp/src/otlp_file_client.cc
index dab2128..6f0e6af 100644
--- a/exporters/otlp/src/otlp_file_client.cc
+++ b/exporters/otlp/src/otlp_file_client.cc
@@ -1234,8 +1234,6 @@ class OPENTELEMETRY_LOCAL_SYMBOL OtlpFileSystemBackend : public OtlpFileAppender
     }
     file_path[file_path_size] = 0;
 
-    std::shared_ptr<FILE> of = std::make_shared<FILE>();
-
     std::string directory_name = FileSystemUtil::DirName(file_path);
     if (!directory_name.empty())
     {
@@ -1290,7 +1288,7 @@ class OPENTELEMETRY_LOCAL_SYMBOL OtlpFileSystemBackend : public OtlpFileAppender
                               << " failed with pattern: " << options_.file_pattern << hint);
       return nullptr;
     }
-    of = std::shared_ptr<std::FILE>(new_file, fclose);
+    std::shared_ptr<FILE> of = std::shared_ptr<std::FILE>(new_file, fclose);
 
     fseek(of.get(), 0, SEEK_END);
     file_->written_size = static_cast<size_t>(ftell(of.get()));
