diff --git a/core/include/prometheus/family.h b/core/include/prometheus/family.h
index 0bc0723..ae156c3 100644
--- a/core/include/prometheus/family.h
+++ b/core/include/prometheus/family.h
@@ -110,7 +110,7 @@ class PROMETHEUS_CPP_CORE_EXPORT Family : public Collectable {
   /// \throw std::runtime_exception on invalid label names.
   template <typename... Args>
   T& Add(const Labels& labels, Args&&... args) {
-    return Add(labels, std::make_unique<T>(args...));
+    return Add(labels, std::unique_ptr<T>(new T(args...)));
   }
 
   /// \brief Remove the given dimensional data.
diff --git a/core/src/registry.cc b/core/src/registry.cc
index 267bf92..8871dca 100644
--- a/core/src/registry.cc
+++ b/core/src/registry.cc
@@ -141,7 +141,7 @@ Family<T>& Registry::Add(const std::string& name, const std::string& help,
     throw std::invalid_argument("Family name already exists");
   }
 
-  auto family = std::make_unique<Family<T>>(name, help, labels);
+  auto family = std::unique_ptr<Family<T>>(new Family<T>(name, help, labels));
   auto& ref = *family;
   families.push_back(std::move(family));
   return ref;
diff --git a/pull/src/endpoint.cc b/pull/src/endpoint.cc
index 5c44bee..b0a3e69 100644
--- a/pull/src/endpoint.cc
+++ b/pull/src/endpoint.cc
@@ -21,7 +21,7 @@ Endpoint::Endpoint(CivetServer& server, std::string uri)
     : server_(server),
       uri_(std::move(uri)),
       endpoint_registry_(std::make_shared<Registry>()),
-      metrics_handler_(std::make_unique<MetricsHandler>(*endpoint_registry_)) {
+      metrics_handler_(std::unique_ptr<MetricsHandler>(new MetricsHandler(*endpoint_registry_))) {
   RegisterCollectable(endpoint_registry_);
   server_.addHandler(uri_, metrics_handler_.get());
 }
@@ -46,7 +46,7 @@ void Endpoint::RegisterAuth(
   // split creating, assigning, and storing to avoid a race-condition when
   // being called the second time and the handler is replaced
   auto new_handler =
-      std::make_unique<BasicAuthHandler>(std::move(authCB), realm);
+      std::unique_ptr<BasicAuthHandler>(new BasicAuthHandler(std::move(authCB), realm));
   server_.addAuthHandler(uri_, new_handler.get());
   auth_handler_ = std::move(new_handler);
 }
diff --git a/pull/src/exposer.cc b/pull/src/exposer.cc
index 38a28e0..cb9fdd2 100644
--- a/pull/src/exposer.cc
+++ b/pull/src/exposer.cc
@@ -20,7 +20,7 @@ Exposer::Exposer(const std::string& bind_address, const std::size_t num_threads,
 
 Exposer::Exposer(std::vector<std::string> options,
                  const CivetCallbacks* callbacks)
-    : server_(std::make_unique<CivetServer>(std::move(options), callbacks)) {}
+    : server_(std::unique_ptr<CivetServer>(new CivetServer(std::move(options), callbacks))) {}
 
 Exposer::~Exposer() = default;
 
@@ -59,7 +59,7 @@ detail::Endpoint& Exposer::GetEndpointForUri(const std::string& uri) {
     return *it->get();
   }
 
-  endpoints_.emplace_back(std::make_unique<detail::Endpoint>(*server_, uri));
+  endpoints_.emplace_back(std::unique_ptr<detail::Endpoint>(new detail::Endpoint(*server_, uri)));
   return *endpoints_.back().get();
 }
 
diff --git a/pull/tests/integration/integration_test.cc b/pull/tests/integration/integration_test.cc
index 96ca5f5..d4306ce 100644
--- a/pull/tests/integration/integration_test.cc
+++ b/pull/tests/integration/integration_test.cc
@@ -22,7 +22,7 @@ using namespace testing;
 class IntegrationTest : public testing::Test {
  public:
   void SetUp() override {
-    exposer_ = std::make_unique<Exposer>("127.0.0.1:0");
+    exposer_ = std::unique_ptr<Exposer>(new Exposer("127.0.0.1:0"));
     auto ports = exposer_->GetListeningPorts();
     base_url_ = std::string("http://127.0.0.1:") + std::to_string(ports.at(0));
   }
diff --git a/push/CMakeLists.txt b/push/CMakeLists.txt
index bc116ed..8111eca 100644
--- a/push/CMakeLists.txt
+++ b/push/CMakeLists.txt
@@ -1,4 +1,9 @@
 
+find_package(zstd QUIET)
+find_package(c-ares QUIET)
+find_package(Libnghttp3 QUIET)
+find_package(Libngtcp2 QUIET)
+find_package(Libnghttp2 QUIET)
 find_package(CURL REQUIRED)
 
 add_library(push
diff --git a/push/src/gateway.cc b/push/src/gateway.cc
index e1ec9d5..095e057 100644
--- a/push/src/gateway.cc
+++ b/push/src/gateway.cc
@@ -20,7 +20,7 @@ namespace prometheus {
 Gateway::Gateway(const std::string& host, const std::string& port,
                  const std::string& jobname, const Labels& labels,
                  const std::string& username, const std::string& password) {
-  curlWrapper_ = std::make_unique<detail::CurlWrapper>(username, password);
+  curlWrapper_ = std::unique_ptr<detail::CurlWrapper>(new detail::CurlWrapper(username, password));
 
   std::stringstream jobUriStream;
   jobUriStream << host << ':' << port << "/metrics/job/" << jobname;
