diff --git a/CMakeLists.txt b/CMakeLists.txt
index b7d6ee8..c76c2da 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -152,6 +152,7 @@ IF(ENABLE_SSL)
         ENDIF()
     ENDIF()
     FIND_PACKAGE(OpenSSL REQUIRED)
+    FIND_PACKAGE(Threads)
     SET(hiredis_ssl_sources
         ssl.c)
     ADD_LIBRARY(hiredis_ssl ${hiredis_ssl_sources})
@@ -171,7 +172,7 @@ IF(ENABLE_SSL)
     ENDIF()
     TARGET_LINK_LIBRARIES(hiredis_ssl PRIVATE OpenSSL::SSL)
     IF(WIN32)
-        TARGET_LINK_LIBRARIES(hiredis_ssl PRIVATE hiredis)
+        TARGET_LINK_LIBRARIES(hiredis_ssl PUBLIC hiredis)
     ENDIF()
     CONFIGURE_FILE(hiredis_ssl.pc.in hiredis_ssl.pc @ONLY)
 
