diff --git a/CMakeLists.txt b/CMakeLists.txt
index b7d6ee8..a9cf559 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -146,12 +146,13 @@ INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/hiredis-config.cmake
 
 
 IF(ENABLE_SSL)
-    IF (NOT OPENSSL_ROOT_DIR)
-        IF (APPLE)
-            SET(OPENSSL_ROOT_DIR "/usr/local/opt/openssl")
-        ENDIF()
-    ENDIF()
+    # IF (NOT OPENSSL_ROOT_DIR)
+    #     IF (APPLE)
+    #         SET(OPENSSL_ROOT_DIR "/usr/local/opt/openssl")
+    #     ENDIF()
+    # ENDIF()
     FIND_PACKAGE(OpenSSL REQUIRED)
+    FIND_PACKAGE(Threads)
     SET(hiredis_ssl_sources
         ssl.c)
     ADD_LIBRARY(hiredis_ssl ${hiredis_ssl_sources})
@@ -171,7 +172,7 @@ IF(ENABLE_SSL)
     ENDIF()
     TARGET_LINK_LIBRARIES(hiredis_ssl PRIVATE OpenSSL::SSL)
     IF(WIN32)
-        TARGET_LINK_LIBRARIES(hiredis_ssl PRIVATE hiredis)
+        TARGET_LINK_LIBRARIES(hiredis_ssl PUBLIC hiredis Bcrypt ws2_32 gdi32 advapi32 crypt32 user32)
     ENDIF()
     CONFIGURE_FILE(hiredis_ssl.pc.in hiredis_ssl.pc @ONLY)
 
