diff --git a/CMakeLists.txt b/CMakeLists.txt
index c5368c08..f508205b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,8 +1,5 @@
 # Use at least CMake 3.3
-cmake_minimum_required (VERSION 3.3.0)
-cmake_policy(VERSION 3.2.2)
-cmake_policy(SET CMP0054 NEW)
-cmake_policy(SET CMP0057 NEW)
+cmake_minimum_required (VERSION 3.16)
 
 # Set up the project
 project (civetweb)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 43f3a77e..15ff20d3 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -56,7 +56,11 @@ if (CIVETWEB_ENABLE_SSL)
     find_package(OpenSSL)
     include_directories(${OPENSSL_INCLUDE_DIR})
     message(STATUS "OpenSSL include directory: ${OPENSSL_INCLUDE_DIR}")
-    target_link_libraries(civetweb-c-library ${OPENSSL_LIBRARIES})
+    if(TARGET OpenSSL::SSL)
+      target_link_libraries(civetweb-c-library OpenSSL::SSL OpenSSL::Crypto)
+    else()
+      target_link_libraries(civetweb-c-library ${OPENSSL_LIBRARIES})
+    endif()
   endif()
 endif()
 
diff --git a/src/civetweb.c b/src/civetweb.c
index 9e321edf..fc3be076 100644
--- a/src/civetweb.c
+++ b/src/civetweb.c
@@ -1775,7 +1775,9 @@ typedef struct SSL_CTX SSL_CTX;
  * the standard memory functions.
  *
  * #define OPENSSL_free free */
+#ifndef free
 #define free free
+#endif
 // disable for boringssl
 #define CONF_modules_unload(a) ((void)0)
 #define ENGINE_cleanup() ((void)0)
