diff --git a/CMakeLists.txt b/CMakeLists.txt
index eabcb90..542d390 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -825,7 +825,7 @@ if (MSVC)
        # Turn off pointless microsoft security warnings.
        add_definitions(-D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE)
        # Fail the build if any warnings
-       add_compile_options(/W3 /WX)
+       # add_compile_options(/W3 /WX)
 endif(MSVC)

 if (MINGW)
